src/rust/lib/graph-editor/src/component/visualization/fullscreen.rs

```diff
@@ -0,0 +1,338 @@
+            }
+            StateModel::TransitioningToFullscreen { target_state, ..} => {
+                let resize_handle = target_state.make_resize_handle();
+                StateModel::Fullscreen { data: target_state, resize_handle }
+            }
+            other    => other,
+        };
+        *self = new_state;
+    }
+
+    /// Returns whether there is a component that is in fullscreen mode. Animation phases count as
+    /// still in fullscreen mode.
+    pub fn is_fullscreen(&self) -> bool {
+        match self {
+            StateModel::NotFullscreen{..} => false,
+            _                             => true,
+        }
+    }
+}
+
+impl<T> Default for StateModel<T> {
+    fn default() -> Self {
+        StateModel::NotFullscreen
+    }
+}
+
+
+
+// =======================
+// === FullscreenState ===
+// =======================
+
+/// The `FullscreenState` manages the state changes between fullscreen mode and non-fullscreen mode
+/// for a UI component. It creates animations for the state changes and ensure that the component
+/// cannot come into an illegal state during the transition.
+#[derive(Debug,CloneRef,Derivative)]
+#[derivative(Clone(bound=""))]
+pub struct FullscreenState<T> {
+    network   : frp::Network,
+    state     : Rc<RefCell<StateModel<T>>>,
+    animation : DynSimulator<f32>,
+}
+
+impl<T:Fullscreenable> Default for FullscreenState<T> {
+    fn default() -> Self {
+        let network    = frp::Network::new();
+        let state      = Rc::new(RefCell::new(StateModel::<T>::default()));
+        let weak_state = Rc::downgrade(&state);
+        let animation  = animation(&network, move |value| {
+            transition_animation_fn(weak_state.clone_ref(), value);
+        });
+        FullscreenState{network,state,animation}
+    }
+}
+
+impl<T:Fullscreenable> FullscreenState<T> {
+    /// Returns whether there is a component that is in fullscreen mode. Animation phases count as
+    /// still in fullscreen mode.
+    pub fn is_fullscreen(&self) -> bool {
+        self.state.borrow().is_fullscreen()
+    }
+
+    /// Enables fullscreen mode for the given component. Does nothing if we are in an animation
+    /// phase, or already in fullscreen mode.
+    pub fn enable_fullscreen(&self, target:T, scene:Scene) {
+        if !self.is_fullscreen() {
+            self.transition_to_fullscreen(target, scene)
+        }
+    }
+
+    /// Disables fullscreen mode for the given component. Does nothing if we are in an animation
+    /// phase, or not in fullscreen mode.
+    pub fn disable_fullscreen(&self) {
+        let fullscreen_data = {
+            let state = self.state.borrow();
+            let state = state.deref();
+            if let StateModel::Fullscreen {data, ..} = state {
+                Some(data.clone())
+            } else {
+                None
+            }
+        };
+        if let Some(data) = fullscreen_data {
+            self.transition_to_non_fullscreen(&data)
+        }
+    }
```

Comment:
If that was done in frp, the implementation would be one line - `fullscreen.set_target(1.0)`.